File: xml-performance/test-drive-xml-sax/src/main/java/org/jesperancinha/test/drive/xml/XmlBookSAXParserManager.java
package org.jesperancinha.test.drive.xml;

import java.io.InputStream;

import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;

import org.jesperancinha.test.drive.sax.books.SAXBooksParser;

public class XmlBookSAXParserManager extends XmlBookParserBuilder {
	private SAXBooksParser handler = null;

	public XmlBookSAXParserManager(InputStream ioStream) {
		super(ioStream);
	}

	@Override
	public String getBookTitle(int bookNumber) {
		return handler.getBooks().get(bookNumber).getTitle();
	}

	@Override
	public int getNumberOfPages(int bookNumber) {
		return handler.getBooks().get(bookNumber).getPages();
	}

	@Override
	public String getGenre(int bookNumber) {
		return handler.getBooks().get(bookNumber).getGenre();
	}

	@Override
	public String getAuthor(int bookNumber) {
		return handler.getBooks().get(bookNumber).getAuthor();
	}

	@Override
	public int getNumberOfBooks() {
		return handler.getBooks().size();
	}

	@Override
	public void init() throws Exception {
		SAXParserFactory spf = SAXParserFactory.newInstance();
		spf.setNamespaceAware(true);
		SAXParser saxParser = spf.newSAXParser();
		XMLReader xmlReader = saxParser.getXMLReader();
		handler = new SAXBooksParser();
		xmlReader.setContentHandler(handler);
		xmlReader.parse(new InputSource(getIoStream()));
	}
}

File: xml-performance/test-drive-xml-sax/src/main/kotlin/org/jesperancinha/test/drive/xml/XmlBookSAXParserManager.kt
package org.jesperancinha.test.drive.xml

import org.jesperancinha.test.drive.sax.books.SAXBooksParser
import org.xml.sax.InputSource
import java.io.InputStream
import javax.xml.parsers.SAXParserFactory

class XmlBookSAXParserManager(ioStream: InputStream) : XmlBookParserBuilder(ioStream) {
    private val handler: SAXBooksParser by lazy { SAXBooksParser() }
    override fun getBookTitle(bookNumber: Int): String? {
        return handler.getBooks()[bookNumber].title
    }

    override fun getNumberOfPages(bookNumber: Int): Int {
        return handler.getBooks()[bookNumber].pages
    }

    override fun getGenre(bookNumber: Int): String? {
        return handler.getBooks()[bookNumber].genre
    }

    override fun getAuthor(bookNumber: Int): String? {
        return handler.getBooks()[bookNumber].author
    }

    override val numberOfBooks: Int by lazy { handler.getBooks().size }

    @Throws(Exception::class)
    override fun init() {
        val spf = SAXParserFactory.newInstance()
        spf.isNamespaceAware = true
        val saxParser = spf.newSAXParser()
        val xmlReader = saxParser.xmlReader
        xmlReader.contentHandler = handler
        xmlReader.parse(InputSource(ioStream))
    }
}